### ISSUE TITLE
ValueError when debugging numpy app

### ISSUE THREAD

**Issue by huguesv** [Microsoft/ptvsd#465]
Create a txt file with 646200 floating point values, one per row.

Create a .py script:
```
import numpy as np
np.loadtxt(yourFileNameHere)
```
Running without debugging or from the interactive prompt succeeds.

Running with the debugger results in the exception:
```ValueError: cannot resize an array that references or is referenced by another array in this way. Use the resize function```

Note: this was filed at https://developercommunity.visualstudio.com/content/problem/264289/numpyloadtxt-fails-while-debugging.html and it didn't repro when I tried it using the experiemental debugger, but when the user tried again with the experimental debugger and was able to reproduce. Maybe someone else here will have different result.

**Comment by karthiknadig** [comment]
Does not repro with the latest master:
![image](https://user-images.githubusercontent.com/3840081/41012730-e7fe2ec8-68f7-11e8-91f1-59b8f768e008.png)

**Comment by karthiknadig** [comment]
Use on developer community reported that they are not able to repro this issue. Re-open if seen again.

**Comment by karthiknadig** [comment]
This stack is from the developer community link:
```
Traceback (most recent call last):
File "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\Extensions\Microsoft\Python\Core\ptvsd_launcher.py", line 111, in
 vspd.debug(filename, port_num, debug_id, debug_options, run_as)
File "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\Extensions\Microsoft\Python\Core\Packages\ptvsd\debugger.py", line 36, in
 debug run(address, filename, args, kwargs) 
File "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\Extensions\Microsoft\Python\Core\Packages\ptvsd_main.py", line 47, in
 run_file run(argv, addr, *kwargs) 
File "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\Extensions\Microsoft\Python\Core\Packages\ptvsd_main.py", line 98, in
 _run _pydevd.main() 
File "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\Extensions\Microsoft\Python\Core\Packages\ptvsd\pydevd\pydevd.py", line 1628, in
 main globals = debugger.run(setup['file'], None, None, is_module)
File "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\Extensions\Microsoft\Python\Core\Packages\ptvsd\pydevd\pydevd.py", line 1035, in
 run pydev_imports.execfile(file, globals, locals) # execute the script
File "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\Extensions\Microsoft\Python\Core\Packages\ptvsd\pydevd_pydev_imps_pydev_execfile.py", line 25, in
 execfile exec(compile(contents+"\n", file, 'exec'), glob, loc)
File "C:\Users\clnorris\Documents\Projects\MyCardio\Pleth_Alg_Eval\PPG\test_PPG.py", line 21, in
 np.loadtxt(txtFile)
File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python36_64\lib\site-packages\numpy\lib\npyio.py", line 1099, in
 loadtxt X.resize(nshape)
ValueError: cannot resize an array that references or is referenced by another array in this way. Use the resize function
```

**Comment by karthiknadig** [comment]
This issue was reported in VSC feedback survey.

**Comment by xenocrates87** [comment]
I have the same issue with a csv file consisting of 10 million lines and 15 columns. When running without debugging it completes without errors. When run in VS Code with debugging I get the same ValueError.
![debugerror](https://user-images.githubusercontent.com/885853/42266726-84b05486-7f77-11e8-8f57-064e97bfdfbc.PNG)

**Comment by karthiknadig** [comment]
@fabioz

**Comment by fabioz** [comment]
I'm investigating this and I the only way I could reproduce it is by stepping into the function with error, although the reports show that it can happen at other times too. 

Given that it's random, this seems to happen when sys._current_frames() is being called, which may temporarily raise a reference to the frame in another thread (it's used on some times in the debugger) and by doing so, apparently python ends up creating a new reference from frame.f_locals (only a method cellvar reference is expected -- on Python 3.6).

Now, unfortunately, I don't see how to fix that in the debugger side... as just peeking may make numpy misbehave and this is essential for the debugger to work. 

The best I can say is that this resize should have `refcheck=False` in order to work properly with the debugger.

Below is a screenshot showing the 2 references during the debugging (when a breakpoint is hit at that context), which will make numpy break.

![numpy_error](https://user-images.githubusercontent.com/117621/42451104-9b92e4f0-835c-11e8-8d18-63b6e5eccaa5.png)

As a note, any debugger which had tracing always on would fail consistently -- i.e.: the code below always fails:

```
import sys

def trace_dispatch(frame, event, arg):
    return trace_dispatch

sys.settrace(trace_dispatch)

def method():
    import numpy
    Y = numpy.array([1] * 10, numpy.float32)
    Y.resize([20, 20])

method()
```

So, to summarize, this bug can be closed as there's no workaround for the debugger.

Possible workarounds for final users include adding a `refcheck=False` in the line that has the problem (could probably be escalated to numpy itself) or fixing Python so that when just getting the frame the `f_locals` is still not computed and is only created on demand as the debugger didn't actually get the locals, it just got the frame, but that was enough to raise the reference count to the array from `frame.f_locals`.
